src/compiler/compiler.rs

```diff
@@ -921,129 +938,166 @@ where
         // This happens when sccache is used as a `RUSTC_WRAPPER` and another tool is used as a
         // `RUSTC_WORKSPACE_WRAPPER`. In that case rustc will be the first argument rather than the command.
         //
         // The check for the `CARGO` env acts as a guardrail against false positives.
         // https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
-        args.iter().next().and_then(|arg1| {
+        if let Some(arg1) = args.iter().next() {
             if is_rustc_like(arg1) {
-                Some(PathBuf::from(arg1))
-            } else {
-                None
+                must_be_rust = true;
+                rustc_executable = PathBuf::from(arg1)
             }
-        })
-    } else {
-        None
+        }
     };
 
-    let creator1 = creator.clone();
     let pool = pool.clone();
-    let cwd = cwd.to_owned();
-    if let Some(rustc_executable) = rustc_executable {
-        // Sanity check that it's really rustc.
-        let mut child = creator.clone().new_command_sync(executable);
-        // We're wrapping rustc if the executable doesn't match the detected rustc_executable. In this case the wrapper
-        // expects rustc as the first argument.
-        if rustc_executable != executable {
-            child.arg(&rustc_executable);
-        }
-
-        child.env_clear().envs(ref_env(env)).args(&["-vV"]);
-
-        let rustc_vv = run_input_output(child, None).await.map(|output| {
-            if let Ok(stdout) = String::from_utf8(output.stdout.clone()) {
-                if stdout.starts_with("rustc ") {
-                    return Ok(stdout);
-                }
+    if !must_be_rust && is_like_c_compiler(&rustc_executable) {
+        let executable = executable.to_owned();
+        let cc = detect_c_compiler(creator, executable, args, env.to_vec(), pool).await;
+        return cc.map(|c| (c, None));
+    }
+
+    // Even if it does not look like rustc like it might still be rustc driver
+    // so we do full check
+    match verify_rust_compiler(
+        creator.clone(),
+        executable,
+        rustc_executable,
+        env,
+        cwd.to_owned(),
+        dist_archive,
+        pool.clone(),
+    )
+    .await
+    {
+        Ok(res) => Ok(res),
+        Err(e) => {
+            if !must_be_rust {
+                let executable = executable.to_owned();
+                let cc = detect_c_compiler(creator, executable, args, env.to_vec(), pool).await;
+                cc.map(|c| (c, None))
+            } else {
+                Err(e)
             }
-            Err(ProcessError(output))
-        })?;
+        }
+    }
+}
 
-        match rustc_vv {
-            Ok(rustc_verbose_version) => {
-                let rustc_executable2 = rustc_executable.clone();
+/// Tries to verify that provided executable is really rust compiler
+/// or rust driver
+async fn verify_rust_compiler<T>(
```

Comment:
Requires more documentation on the disriminator used and the function name needs a better name.

Reply:
What is wrong with name? Main part of this function is:
```rust
// rustc -vV verification status
match rustc_vv {
   ...
```
anything specific you have in mind?

I also do not think there is much to document about discriminator used as there is no discriminator used. We check for rust compiler using -vV instead of using binary name and then verifying compiler with -vV.

Reply:
It does attempt to _resolve_ a given path to the true (`rustc`) compiler binary with a failure mode if it cannot, so verification is a red hering and doesn't describe what it is doing. That's my main issue

Reply:
What about `resolve_rust_compiler`?